ddtrace/contrib/logging/patch.py

```diff
@@ -37,14 +37,15 @@ class DDLogRecord(object):
 
 def _get_current_span(tracer=None):
     """Helper to get the currently active span"""
-    if not tracer:
-        tracer = ddtrace.tracer
 
     # We might be calling this during library initialization, in which case `ddtrace.tracer` might
     # be the `tracer` module and not the global tracer instance.
     if not getattr(tracer, "enabled", False):
         return None
 
+    if not tracer:
```

Comment:
Reordering this logic is needed to handle the scenario where `ddtrace-run` is used with `DD_LOGS_INJECTION=true DD_TRACE_DEBUG=true DD_CALL_BASIC_CONFIG=true` when there is a `ddtrace` logger. 

If a tracer object isn't available, it doesn't mean that `ddtrace.tracer` is available. If this is checked first, it can lead to an error that `ddtracer.tracer` doesn't exist:

```
E           File "/root/project/.riot/venv_py3103/bin/ddtrace-run", line 33, in <mod...eError: partially initialized module \'ddtrace\' has no attribute \'tracer\' (most likely due to a circular import)
```

This is because `tracer` doesn't exist under `ddtrace`:
```
dir(ddtrace)
['Pin', 'Span', '__builtins__', '__cached__', '__doc__', '__file__', '__loader__', '__name__', '__package__', '__path__', '__spec__', '_hooks', '_logger', '_monkey', 'config', 'configure_ddtrace_logger', 'constants', 'context', 'contrib', 'ext', 'filters', 'internal', 'patch', 'patch_all', 'pin', 'propagation', 'provider', 'sampler', 'settings', 'span', 'vendor', 'version']

```

Alternatively, I could also just check if `tracer` exists under `ddtrace` with `getattr`, but I'm not sure if there's a specific reason to check the tracer this way.

Reply:
If I keep the import as it was previously, that is:
```def _get_current_span(tracer=None):
    """Helper to get the currently active span"""

    if not tracer:
        tracer = ddtrace.tracer

    # We might be calling this during library initialization, in which case `ddtrace.tracer` might
    # be the `tracer` module and not the global tracer instance.
    if not getattr(tracer, "enabled", False):
        return None

    return tracer.current_span()
```


The failing test is `test_ddtrace_run_startup_logging_injection` in `test_integration.py`, where it doesn't know what `ddtrace.tracer` means: https://github.com/DataDog/dd-trace-py/blob/1.x/tests/integration/test_integration.py#L743 .

If the code is unchanged, but `DD_CALL_BASIC_CONFIG=false`, then the logs that get emitted are the ones under `ddtrace.commands.ddtrace_run` (it also doesn't have any trace injection configured, there's no dd.span_id in the logs below) so it looks like these are called before `ddtrace.tracer` exists:

```
b"DEBUG:ddtrace.commands.ddtrace_run:ddtrace root: /<>/dd-trace-py/ddtrace
DEBUG:ddtrace.commands.ddtrace_run:ddtrace bootstrap: /<>/dd-trace-py/ddtrace/bootstrap
DEBUG:ddtrace.commands.ddtrace_run:PYTHONPATH: /<>/dd-trace-py/ddtrace/bootstrap::/<>/dd-trace-py:/<>/dd-trace-py/.riot/venv_py399...site-packages:/<>/dd-trace-py/.riot/venv_py399/lib/python3.9/site-packages
DEBUG:ddtrace.commands.ddtrace_run:sys.path: [list of paths]
DEBUG:ddtrace.commands.ddtrace_run:program executable: /<>/dd-trace-py/.riot/venv_py399/bin/python
"
```

Reply:
As an update:
I tried out removing `asbool` from `_logger.py` and this continues to happen.

This error also happens if these are the env vars used: `DD_LOGS_INJECTION=false DD_TRACE_DEBUG=true DD_CALL_BASIC_CONFIG=true`, meaning the problem is on the log injection setting, not call basic config or `_logger.py`.

When log injection is set, **tracer.py** calls `patch(logging=true)`: https://github.com/DataDog/dd-trace-py/blob/1.x/ddtrace/tracer.py#L79 . 

`logging/patch.py` calls on `ddtrace.tracer` if there's no `tracer` yet: https://github.com/DataDog/dd-trace-py/blob/1.x/ddtrace/contrib/logging/patch.py#L41 .

However, at that point, there isn't a `tracer` yet, which explains my earlier results:

```
dir(ddtrace)
['Pin', 'Span', '__builtins__', '__cached__', '__doc__', '__file__', '__loader__', '__name__', '__package__', '__path__', '__spec__', '_hooks', '_logger', '_monkey', 'config', 'configure_ddtrace_logger', 'constants', 'context', 'contrib', 'ext', 'filters', 'internal', 'patch', 'patch_all', 'pin', 'propagation', 'provider', 'sampler', 'settings', 'span', 'vendor', 'version']

```

As a result, I get the circular import error because there's no `ddtrace.tracer`.

Reply:
The only two lines added to `init` to change this behavior from the current way the tracer is configured: https://github.com/DataDog/dd-trace-py/blob/1.x/ddtrace/__init__.py, would be 

```
from ._logger import configure_ddtrace_logger


configure_ddtrace_logger()
```

🤔

Reply:
Marking as resolved.